### ISSUE TITLE
struct.Struct inheritance with Python 3.12.0

### ISSUE THREAD

**Issue by livrrr** [python/cpython#112358]
# Bug report

### Bug description:

```python
import struct

class MyStruct(struct.Struct):
    def __init__(self):
        super().__init__('>h')

obj = MyStruct()

```
When I run this code I receive an error:
```
Traceback (most recent call last):
  File "/home/user/bug.py", line 7, in <module>
    obj = MyStruct()
          ^^^^^^^^^^
TypeError: Struct() missing required argument 'format' (pos 1)
```
It is rather strange error, I have passed `format` parameter to base class constructor, I receive this error with any value of `format` parameter.
There are no any problems with this code in Python 3.11 and older.

### CPython versions tested on:

3.12

### Operating systems tested on:

Linux

### Linked PRs
* gh-112424
* gh-112426

**Comment by Eclips4** [comment]
bisected to c8c0afc7137ab9f22bf59d591084948ca967c97c
cc @kumaraditya303

**Comment by AlexWaygood** [comment]
Looks like another user also hit this problem, but reported it on the PR rather than opening an issue: https://github.com/python/cpython/pull/94532#issuecomment-1760331238

**Comment by mdickinson** [comment]
Hmm, this isn't nice. I'm not seeing any good options here other than simply reverting the offending commit (and re-opening the issues that PR #94532 was supposed to close). @kumaraditya303?

**Comment by mdickinson** [comment]
> Hmm, this isn't nice. [...]

To elaborate: the existence of `struct.Struct` subclasses that extend the `__init__` method means that `Struct.__new__` can make essentially no assumptions about the arguments it gets. It might get no arguments; it might get just the format, it might get the format plus additional things, or it might get a collection of arguments that doesn't include the format at all. Previously all those things worked because `Struct.__new__` was simply `object.__new__`, which accepted any combination of args and kwargs.

This makes the solution in #94532 of initialising the struct in `Struct.__new__` non-viable - at least, not without backwards compatibility breakage, which would have to be managed in the usual way (deprecation period, etc.).

I've opened #112424 for reversion. Once merged, it'll need to be backported to 3.12, and issues #75960 and #78724 will need to be re-opened.

**Comment by mdickinson** [comment]
Fixed in #112424 (main), #112426 (3.12).

**Comment by mdickinson** [comment]
@serhiy-storchaka: Maybe. I'd failed to recognise the scope of the breakage introduced by #94532, and I think reverting in the short term is the right solution. Longer term, if we can find a properly deprecated path to moving the initialisation to `__new__`, I agree that that would be worth considering.

For the particular issues that #94532 was addressing, I think there are likely less disruptive fixes available.
